File: 367. Valid Perfect Square/IsPerfectSquare.py
'''
https://leetcode.com/problems/valid-perfect-square/
'''


class Solution:
    def isPerfectSquare(self, num: int) -> bool:
        def binary_search(start, end):
            mid = (start + end) //2
            if square(mid) == num:    return True
            if start == mid and square(mid) != num:    return False
            if square(mid) < num:
                return binary_search(mid, end)
            else:
                return binary_search(start, mid)
            
        def square(x):
            return x * x
        
        return binary_search(0, num+1)

File: 367. Valid Perfect Square/is_perfect_square.rs
impl Solution {
    
    // O(logn) time,
    // O(1) space,
    // Approach: binary search, math
    pub fn is_perfect_square(num: i32) -> bool {
        let mut low = 1;
        let mut hi = (46341); // 46,341 is the maximum root u can have for a i32 type
        
        while low <= hi {
            let mid = low + (hi-low)/2;
            let square = (mid*mid);
            
            if square == num {
                return true;
            }
            else if square < num{
                low = mid+1;
            }
            else {
                hi = mid-1;
            }
        }
        
        return false;
    }
}